### ISSUE TITLE
OSError: Illegal byte sequence

### ISSUE THREAD

**Issue by garfieldnate** [tatuylonen/wiktextract#145]
If you write the page files using `--pages-dir pages`, some of the page names may not be allowed file names.  On my OSX (12.3), this lead to the following error:

    OSError: [Errno 92] Illegal byte sequence: 'pages/Words/\U0001f77f'

This was also while parsing pages from the Simple English Wiktionary, so it's possible there are no English Wiktionary pages that cause this error.

**Comment by kristian-clausal** [comment]
That \u0001f77f looks malformed. I think \u is just ASCII code-point #1 (after #0), so ^A 'start of header', which is obviously ancient teleprompter printer stuff that might or might not be relevant in unicode, and if you try out the whole sequence out in a python string:
´´´
>>> print("\u0001f77f")
f77f
>>> print("\u1f77f")
ίf
´´´

But considering it's pages/Words/, I would expect that to maybe be a Hanzi character or something else... but the closest guess 1f77f is an apparently unused point in the Alchemical Symbols block.

Tatu probably has a better idea. It might have to do with how the code generates the directory string.

**Comment by kristian-clausal** [comment]
Tatu told me a quick way to patch over the issue, so I've pushed a commit that handles OSErrors when writing the word page files.

However, we're still not exactly sure what the issue actually is; our suspicion is that OSX handles Unicode a bit differently.

This fix bypasses the issue, so it actually doesn't generate the affected pages correctly, but at least wiktwords should now function. Because this is, from our perspective, a pretty major error, there is a print() for the OSError + word + file name, but hopefully it will appear only once. If the whole output of the script is thousands of lines of this error message, something probably needs to be done, but we don't have any Macs here at the office so it's hard to test.

**Comment by kristian-clausal** [comment]
Reopening for comments; apparently "fix #XXX" in commit messages does close an issue.

**Comment by garfieldnate** [comment]
I pulled the latest and dug a bit deeper.

The try/catch [here](https://github.com/tatuylonen/wikitextprocessor/blob/main/wikitextprocessor/core.py#L66) actually does catch this issue already, but, as I noted elsewhere, for some reason the call to `traceback.format_exception` doesn't print the entire stacktrace, so the actual cause is lost.

Once I disabled that try/catch, I found that the error actually occurs [here](https://github.com/tatuylonen/wiktextract/blob/master/wiktwords#L78), just outside where you added the try/catch. I added the try/catch below:


        try:
            os.makedirs(dirpath, exist_ok=True)
        except OSError as err:
            print("OSError: {}, "
                  "when creating dir name {!r}, for "
                  "title: {!r}".format(err, dirpath, orig_title))

Then got this printout:

    OSError: [Errno 92] Illegal byte sequence: 'pages/Words/\U0001f77f', when creating dir name 'pages/Words/\U0001f77f', for title: '\U0001f77f'

The EN Wiktionary page is [here](https://en.wiktionary.org/wiki/%F0%9F%9D%BD). This codepoint belongs to the [alchemical symbols block](https://www.unicode.org/charts/PDF/U1F700.pdf), and according to all documents I can find is officially not assigned in Unicode. When you do `mkdir 🝽` on OSX you get this error; Linux does not give this error, instead simply creating the directory with the given codepoint for a name. It's perfectly fine to create a directory like `元` on OSX, so Unicode is fine, you just can't use unassigned codepoints.

**Comment by kristian-clausal** [comment]
Good catch with the os.makedirs(). Was that requested Unicode-point the only error of this kind that came up? If it was, then I think we can ignore it by printing an error. I've moved the os.makedirs() inside the try/except block with the file write, I don't think we need to check for the same exception twice (and they should logically both fail at the same time, but just in case).

**Comment by garfieldnate** [comment]
>I don't think we need to check for the same exception twice

Yeah, the two error cases always happened together.

>Was that requested Unicode-point the only error of this kind that came up?

It's the only one I saw consistently, so I think it's safe to say so for now.

>If it was, then I think we can ignore it by printing an error.

Does missing this page  prevent the data on the page from being added to the final output, or does it only prevent the page file (for debugging, inspection, etc. purposes) from being printed?

**Comment by kristian-clausal** [comment]
> Does missing this page prevent the data on the page from being added to the final output, or does it only prevent the page file (for debugging, inspection, etc. purposes) from being printed?

Tatu didn't seem to be worried about it, so I assume there will be no issue. Additionally the cache file, if the offending data is written there, has no problem with file name encoding. Just in case I'll ask him, though.

**Comment by kristian-clausal** [comment]
Yeah, confirmed with Tatu and there shouldn't be any problems. The files being written are independent and nothing depends on any 'memory' from wiktwords that would create phantom files in an index or list of some sort, everything after wiktwords depends on either reading pages-directory files or the cache. Closing as fixed. If something pops up, please feel free to open a new issue.
